packages/server/src/constraints.ts

```diff
@@ -0,0 +1,114 @@
+import { DAY, HOUR, cborEncode, check } from '@atproto/common'
+import * as plc from '@did-plc/lib'
+import { ServerError } from './error'
+import { parseDidKey } from '@atproto/crypto'
+
+export function assertValidIncomingOp(
+  op: unknown,
+): asserts op is plc.OpOrTombstone {
+  const byteLength = cborEncode(op).byteLength
+  if (byteLength > 4000) {
+    throw new ServerError(400, 'Operation too large')
+  }
+  if (!check.is(op, plc.def.opOrTombstone)) {
+    throw new ServerError(400, `Not a valid operation: ${JSON.stringify(op)}`)
+  }
+  if (op.type === 'plc_tombstone') {
+    return
+  }
+  if (op.alsoKnownAs.length > 10) {
+    throw new ServerError(400, 'To many alsoKnownAs items (max 10)')
+  }
+  const akaDupe: Record<string, boolean> = {}
+  for (const aka of op.alsoKnownAs) {
+    if (aka.length > 256) {
+      throw new ServerError(400, `alsoKnownAs field too long (max 256): ${aka}`)
+    }
+    if (akaDupe[aka]) {
+      throw new ServerError(400, `duplicate alsoKnownAs field: ${aka}`)
+    } else {
+      akaDupe[aka] = true
+    }
+  }
+  if (op.rotationKeys.length > 5) {
+    throw new ServerError(400, 'To many rotationKey items (max 5)')
+  }
+  for (const key of op.rotationKeys) {
+    try {
+      parseDidKey(key)
+    } catch (err) {
+      throw new ServerError(400, `Invalid rotationKey: ${key}`)
+    }
+  }
+  const serviceEntries = Object.entries(op.services)
+  if (serviceEntries.length > 10) {
```

Comment:
I'd pull all these constants out, specifically because you have them in the code and then immediately below in the exception message, and they could get out of sync easily.

Reply:
yeah good call 👍